lib/pdcleaner/pdcleaner.go

```diff
@@ -0,0 +1,308 @@
+					if err != nil {
+						// Don't return if cleaning up a deal results in error. Try them all.
+						log.Errorf("cleaning up boost deal %s for piece %s: %s", d.DealUuid.String(), d.ClientDealProposal.Proposal.PieceCID.String(), err.Error())
+					}
+				}
+			}
+		}
+	}
+
+	// Clean up completed/slashed legacy deals
+	for _, d := range legacyDeals {
+		// Confirm deal did not reach termination before Publishing. Otherwise, no need to clean up
+		if d.DealID > abi.DealID(0) {
+			// If deal exists online
+			md, ok := deals[strconv.FormatInt(int64(d.DealID), 10)]
+			if ok {
+				// If deal is slashed or end epoch has passed. No other reason for deal to reach termination
+				toCheck := termOrSlash(md.Proposal.EndEpoch, md.State.SlashEpoch)
+				if toCheck < head.Height() {
+					err = p.pd.RemoveDealForPiece(p.ctx, d.ClientDealProposal.Proposal.PieceCID, d.ProposalCid.String())
+					if err != nil {
+						// Don't return if cleaning up a deal results in error. Try them all.
+						log.Errorf("cleaning up legacy deal %s for piece %s: %s", d.ProposalCid.String(), d.ClientDealProposal.Proposal.PieceCID.String(), err.Error())
+					}
+				}
+			}
+		}
+	}
+
+	// Clean up direct deals if there are any otherwise skip this step
+	if len(completeDirectDeals) > 0 {
+		claims, err := p.full.StateGetClaims(p.ctx, p.miner, tskey)
+		if err != nil {
+			return fmt.Errorf("getting claims for the miner %s: %w", p.miner, err)
+		}
+		// Loading miner actor locally is preferred to avoid getting unnecessary data from full.StateMinerActiveSectors()
+		mActor, err := p.full.StateGetActor(p.ctx, p.miner, tskey)
+		if err != nil {
+			return fmt.Errorf("getting actor for the miner %s: %w", p.miner, err)
+		}
+		store := adt.WrapStore(p.ctx, cbor.NewCborStore(blockstore.NewAPIBlockstore(p.full)))
+		mas, err := miner.Load(store, mActor)
+		if err != nil {
+			return fmt.Errorf("loading miner actor state %s: %w", p.miner, err)
+		}
+		activeSectors, err := miner.AllPartSectors(mas, miner.Partition.ActiveSectors)
+		if err != nil {
+			return fmt.Errorf("getting active sector sets for miner %s: %w", p.miner, err)
+		}
+
+		for _, d := range completeDirectDeals {
+			// AllocationID and ClaimID should match
+			cID := verifregtypes.ClaimId(d.AllocationID)
+			c, ok := claims[cID]
+			if ok {
+				present, err := activeSectors.IsSet(uint64(c.Sector))
+				if err != nil {
+					return fmt.Errorf("checking if bitfield is set: %w", err)
+				}
+				// Each claim is created with ProveCommit message. So, a sector in claim cannot be unproven.
+				// it must be either Active(Proving, Faulty, Recovering) or terminated. If bitfield is not set
+				// then sector must have been terminated. This method will also account for future change in sector numbers
+				// of a claim. Even if the sector is changed then it must be Active as this change will require a
+				// ProveCommit message
+				if !present {
+					err = p.pd.RemoveDealForPiece(p.ctx, d.PieceCID, d.ID.String())
```

Comment:
I'm not clear on the sectors bitfield bit but the rest of this seems to me to be saying:

1. Find a list of "completed" direct deals (what's the criteria for "completed", does that mean that the deal has been handed off to the miner?)
2. Look up the allocation ID in the list of claims to see if it's been converted to a claim
3. Check that the sector has been activated, or at least is in an active state
4. if it's not active, then clean up the deal from the local piece directory

So, what are the cases this is trying to account for? There's two that I think I see:

1. "got the claim, but failed to activate the sector at the last hurdle"—which should be quite hard, but looking at the code not entirely impossible.
2. The sector was terminated but claim is dangling, as outlined in https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0045.md#sector-termination, which kind of sucks but I guess this is what makes the two-step process necessary?

Reply:
Sector bitfield comes from Actor state and account for all the sectors which are active for a miner. Active sectors include the ones which are not terminated or are not yet on-chain. 
1. Complete means that the sector it AP()ed in is now in proving state.

Accounting:
1. Any deals for which we don't activate a sector. We should clean up those.
2. Any deals which were claimed but now claim has expired and sector has been terminated should be cleaned up.
3. Any deal where sector has changed should not be cleaned up. (Future reseal).
4. Any deal which creates a claim but sector goes to faulty state instead of proving should not be cleaned up. This can work as faulty sectors are part of Active sector bitfield.

2 more things to test:
1. How does Boost behaves when sector reaching faulty state instead of proving.
2. Confirm if newly proving sector (without first wdPost) are part of the Active Bitfield. If not then update code to account for those. 


To sum it up, if I am not holding a deal, I should not hold the metadata either. We can expand this logic in future if more cases or edge cases are found.

Reply:
sounds good and reasonable